File: character_logic/librang.py
"""Librang battle and AI hooks."""

from __future__ import annotations

from typing import Any

from .common import floor_int


CHARACTER_ID = "librang"
BALANCE = "균형"
PRAYER = "평형의 기도"
HIDDEN_COUNTERS = {PRAYER}
JUDGMENT_BASE_PERCENT = 25
JUDGMENT_RESERVE_MP = 50


def init_unique_state(fighter: Any, unique_names: set[str]) -> None:
    if BALANCE in unique_names:
        fighter.counters[BALANCE] = 0


def counter_state_text(fighter: Any, name: str, value: Any) -> str | None:
    if name == BALANCE:
        return f"{BALANCE} {value}"
    return None


def extra_state_parts(battle: Any, fighter: Any) -> list[str]:
    if int(fighter.counters.get(PRAYER, 0)) <= 0:
        return []
    mult = 1 + _balance(fighter) * 0.2
    return [f"{PRAYER}: 공격 피해 x{mult:g}"]


def needs_battle_log(fighter: Any) -> bool:
    return True


def render_battle_log(battle: Any, fighter: Any, lines: list[str]) -> None:
    opponent = battle.opponent(fighter)
    own_attacks, own_non_attacks = _selected_counts(battle, fighter)
    opp_attacks, opp_non_attacks = _selected_counts(battle, opponent)
    lines.append(f"자신 선택: 공격 {own_attacks} / 비공격 {own_non_attacks}")
    lines.append(f"상대 선택: 공격 {opp_attacks} / 비공격 {opp_non_attacks}")


def counter_resource_value(fighter: Any, name: str, raw: Any) -> float | None:
    if name == BALANCE and isinstance(raw, int):
        return raw * 240
    if name == PRAYER:
        return 180 if int(raw or 0) > 0 else 0.0
    return None


def modify_cost(battle: Any, fighter: Any, action: Any, cost: int) -> int:
    if action.is_skill(CHARACTER_ID, 1):
        cost -= _balance(fighter)
    return cost


def apply_condition_effects(battle: Any, choice: Any) -> bool | None:
    actor = choice.actor
    action = choice.action
    if action.is_skill(CHARACTER_ID, 0):
        bonus = _balance(actor) * 2
        if bonus > 0:
            choice.power = (choice.power or 0) + bonus
            print(f"{BALANCE} 중첩 수 {_balance(actor)}로 위력이 {bonus} 증가했다.")
    if action.is_skill(CHARACTER_ID, 2) and _balance(actor) < 1:
        return False
    return None


def estimated_power(battle: Any, actor: Any, target: Any, action: Any, power: int) -> int:
    if action.is_skill(CHARACTER_ID, 0):
        power += _balance(actor) * 2
    return power


def attack_damage_multipliers(battle: Any, choice: Any) -> list[float]:
    if choice.action.is_attack and int(choice.actor.counters.get(PRAYER, 0)) > 0:
        return [1 + _balance(choice.actor) * 0.2]
    return []


def estimated_damage_multipliers(battle: Any, actor: Any, target: Any, action: Any) -> list[float]:
    if action.is_attack and int(actor.counters.get(PRAYER, 0)) > 0:
        return [1 + _balance(actor) * 0.2]
    return []


def on_hit_after_defense(battle: Any, choice: Any, total_damage: int) -> None:
    actor = choice.actor
    target = battle.opponent(actor)
    action = choice.action
    if action.is_skill(CHARACTER_ID, 0):
        attacks, non_attacks = _selected_counts(battle, target)
        extra = max(0, non_attacks - attacks)
        battle.fixed_damage(target, extra, action.name)
    elif action.is_skill(CHARACTER_ID, 3):
        battle.fixed_damage(target, _judgment_fixed_damage(target, _balance(actor)), action.name)


def apply_non_attack_effects(battle: Any, choice: Any) -> bool:
    actor = choice.actor
    action = choice.action
    if action.is_skill(CHARACTER_ID, 1):
        battle.apply_defense(actor, action.name)
        return True
    if action.is_skill(CHARACTER_ID, 2):
        actor.counters[PRAYER] = 2
        mult = 1 + _balance(actor) * 0.2
        print(f"다음 턴 동안 공격 피해가 x{mult:g}가 된다.")
        return True
    return False


def on_defense_hit(battle: Any, choice: Any, total_damage: int) -> None:
    attacker = choice.actor
    defender = battle.opponent(attacker)
    if defender.defense_name != "지킨다는 것의 무거움":
        return
    attacks, non_attacks = _selected_counts(battle, attacker)
    extra = max(0, attacks - non_attacks)
    battle.fixed_damage(attacker, extra, defender.defense_name)


def on_turn_end(battle: Any, fighter: Any) -> None:
    if battle.turn % 2 != 0:
        return
    attacks, non_attacks = _selected_counts(battle, fighter)
    if attacks == non_attacks:
        battle.add_counter(fighter, BALANCE, 1)


def decrement_counters(fighter: Any) -> None:
    if PRAYER not in fighter.counters:
        return
    remaining = int(fighter.counters.get(PRAYER, 0))
    if remaining <= 1:
        fighter.counters.pop(PRAYER, None)
    else:
        fighter.counters[PRAYER] = remaining - 1


def modify_fixed_damage_to_opponent(battle: Any, actor: Any, target: Any, amount: int) -> int:
    bonus = _balance(actor)
    if bonus <= 0 or amount <= 0:
        return amount
    print(f"냉혹한 심판자: 고정 피해가 {bonus} 증가했다.")
    return amount + bonus


def setup_value(battle: Any, actor: Any, target: Any, action: Any) -> float:
    balance = _balance(actor)
    if action.is_skill(CHARACTER_ID, 1):
        incoming = battle.estimate_best_incoming_damage(target, actor)
        attacks, non_attacks = _selected_counts(battle, target)
        punish = max(0, attacks - non_attacks)
        return incoming * 0.25 + punish * 150 + balance * 60
    if action.is_skill(CHARACTER_ID, 2) and balance >= 1:
        if int(actor.counters.get(PRAYER, 0)) > 0:
            return 0.0
        return 560 + balance * 300
    return 0.0


def ai_score(
    battle: Any,
    actor: Any,
    target: Any,
    action: Any,
    expected_damage: float,
    hit_rate: float,
) -> float:
    value = 0.0
    balance = _balance(actor)
    own_attacks, own_non_attacks = _selected_counts(battle, actor)
    target_attacks, target_non_attacks = _selected_counts(battle, target)
    incoming = battle.estimate_best_incoming_damage(target, actor)

    value += _balance_timing_value(battle, actor, action, own_attacks, own_non_attacks)

    if action.is_skill(CHARACTER_ID, 0):
        extra = max(0, target_non_attacks - target_attacks)
        value += balance * 180 + extra * 260 * hit_rate
        if extra > 0:
            value += 360

    elif action.is_skill(CHARACTER_ID, 1):
        punish = max(0, target_attacks - target_non_attacks)
        incoming_weight = 0.35
        if target_attacks > target_non_attacks or incoming >= actor.hp * 0.75:
            incoming_weight = 0.95
        value += incoming * incoming_weight + punish * 330
        if incoming >= actor.hp:
            value += 2100
        if battle.turn % 2 == 0 and _projected_equal(actor, action, own_attacks, own_non_attacks):
            value += 460

    elif action.is_skill(CHARACTER_ID, 2):
        if balance < 1:
            value -= 3200
        elif int(actor.counters.get(PRAYER, 0)) > 0:
            value -= 900
        else:
            value += 780 + balance * 460
            if actor.mp >= JUDGMENT_RESERVE_MP + action.mp // 2:
                value += 520

    elif action.is_skill(CHARACTER_ID, 3):
        fixed = _judgment_fixed_damage(target, balance) + balance
        value += fixed * 2.8 * hit_rate
        if expected_damage + fixed * hit_rate >= target.hp:
            value += 7200
        elif actor.mp < battle.effective_cost(actor, action) + 18:
            value -= 520
        if int(actor.counters.get(PRAYER, 0)) > 0:
            value += expected_damage * 0.85

    elif action.is_common_action("meditation"):
        if actor.mp < 50:
            value += 360 + max(0, 64 - actor.mp) * 9
        if battle.turn % 2 == 0 and _projected_equal(actor, action, own_attacks, own_non_attacks):
            value += 520
        if actor.mp >= 92:
            value -= 420

    elif action.is_common_action("normal_attack"):
        if balance >= 2 and int(actor.counters.get(PRAYER, 0)) > 0:
            value += expected_damage * 0.7

    return value


def would_condition_fail(battle: Any, actor: Any, target: Any, action: Any) -> bool | None:
    if action.is_skill(CHARACTER_ID, 2):
        return _balance(actor) < 1
    return None


def _balance(fighter: Any) -> int:
    return int(fighter.counters.get(BALANCE, 0))


def _judgment_fixed_damage(target: Any, balance: int) -> int:
    return floor_int(target.hp * (JUDGMENT_BASE_PERCENT + balance * 5) / 100)


def _selected_counts(battle: Any, fighter: Any) -> tuple[int, int]:
    attacks = 0
    non_attacks = 0
    for action_key in fighter.selected_history:
        if battle.action_key_is_attack(fighter, action_key):
            attacks += 1
        else:
            non_attacks += 1
    return attacks, non_attacks


def _projected_counts(action: Any, attacks: int, non_attacks: int) -> tuple[int, int]:
    if action.is_attack:
        return attacks + 1, non_attacks
    return attacks, non_attacks + 1


def _projected_equal(actor: Any, action: Any, attacks: int, non_attacks: int) -> bool:
    projected_attacks, projected_non_attacks = _projected_counts(action, attacks, non_attacks)
    return projected_attacks == projected_non_attacks


def _balance_timing_value(
    battle: Any,
    actor: Any,
    action: Any,
    attacks: int,
    non_attacks: int,
) -> float:
    projected_attacks, projected_non_attacks = _projected_counts(action, attacks, non_attacks)
    current_gap = abs(attacks - non_attacks)
    projected_gap = abs(projected_attacks - projected_non_attacks)
    if current_gap >= 2:
        if projected_gap < current_gap:
            return 920 + (current_gap - projected_gap) * 260
        return -min(1200, (projected_gap - current_gap + 1) * 360)
    if battle.turn % 2 == 0:
        if projected_gap == 0:
            return 760
        if projected_gap <= 2:
            return 640 - projected_gap * 50
        return -min(900, projected_gap * 240)
    if projected_gap < current_gap:
        return 340
    if projected_gap <= 2:
        return 150 - projected_gap * 45
    return -260

File: web/battle-engine/character-logic/librang.js
"use strict";

const CHARACTER_ID = "librang";
const BALANCE = "균형";
const PRAYER = "평형의 기도";
const JUDGMENT_BASE_PERCENT = 25;
const JUDGMENT_RESERVE_MP = 50;

function floorInt(value) {
  return Math.floor(value);
}

function balance(fighter) {
  return Math.max(0, Number(fighter.counters[BALANCE] || 0));
}

function actionKeyIsAttack(fighter, key) {
  if (key === "common:normal_attack") return true;
  if (String(key).startsWith("common:")) return false;
  const [id, slotText] = String(key).split(":");
  const slot = Number(slotText);
  return id === fighter.characterId && Number.isInteger(slot) && fighter.data.skills?.[slot]?.power != null;
}

function selectedCounts(fighter) {
  let attacks = 0;
  let nonAttacks = 0;
  for (const key of fighter.selectedHistory) {
    if (actionKeyIsAttack(fighter, key)) attacks += 1;
    else nonAttacks += 1;
  }
  return [attacks, nonAttacks];
}

function addCounter(battle, fighter, name, amount) {
  if (typeof battle.addCounter === "function") {
    battle.addCounter(fighter, name, amount);
    return;
  }
  const before = Number(fighter.counters[name] || 0);
  fighter.counters[name] = before + Number(amount);
  battle.logs.push(`${fighter.name}의 ${name} ${before} -> ${fighter.counters[name]}`);
}

function judgmentFixedDamage(target, stacks) {
  return floorInt(target.hp * (JUDGMENT_BASE_PERCENT + stacks * 5) / 100);
}

function projectedCounts(action, attacks, nonAttacks) {
  return action.isAttack ? [attacks + 1, nonAttacks] : [attacks, nonAttacks + 1];
}

function projectedEqual(action, attacks, nonAttacks) {
  const [projectedAttacks, projectedNonAttacks] = projectedCounts(action, attacks, nonAttacks);
  return projectedAttacks === projectedNonAttacks;
}

function balanceTimingValue(battle, action, attacks, nonAttacks) {
  const [projectedAttacks, projectedNonAttacks] = projectedCounts(action, attacks, nonAttacks);
  const currentGap = Math.abs(attacks - nonAttacks);
  const projectedGap = Math.abs(projectedAttacks - projectedNonAttacks);
  if (currentGap >= 2) {
    if (projectedGap < currentGap) return 920 + (currentGap - projectedGap) * 260;
    return -Math.min(1200, (projectedGap - currentGap + 1) * 360);
  }
  if (battle.turn % 2 === 0) {
    if (projectedGap === 0) return 760;
    if (projectedGap <= 2) return 640 - projectedGap * 50;
    return -Math.min(900, projectedGap * 240);
  }
  if (projectedGap < currentGap) return 340;
  if (projectedGap <= 2) return 150 - projectedGap * 45;
  return -260;
}

module.exports = {
  HIDDEN_COUNTERS: new Set([PRAYER]),

  initUniqueState(fighter, uniqueNames) {
    if (uniqueNames.has(BALANCE)) fighter.counters[BALANCE] = 0;
  },

  counterStateText(_fighter, name, value) {
    return name === BALANCE ? `${BALANCE} ${value}` : undefined;
  },

  extraStateParts(_battle, fighter) {
    if (Number(fighter.counters[PRAYER] || 0) <= 0) return [];
    return [`${PRAYER}: 공격 피해 x${1 + balance(fighter) * 0.2}`];
  },

  needsBattleLog() {
    return true;
  },

  renderBattleLog(battle, fighter, lines) {
    const opponent = battle.opponent(fighter);
    const [ownAttacks, ownNonAttacks] = selectedCounts(fighter);
    const [oppAttacks, oppNonAttacks] = selectedCounts(opponent);
    lines.push(`자신 선택: 공격 ${ownAttacks} / 비공격 ${ownNonAttacks}`);
    lines.push(`상대 선택: 공격 ${oppAttacks} / 비공격 ${oppNonAttacks}`);
  },

  counterResourceValue(_fighter, name, raw) {
    if (name === BALANCE && Number.isInteger(raw)) return raw * 240;
    if (name === PRAYER) return Number(raw || 0) > 0 ? 180 : 0;
    return undefined;
  },

  modifyCost(_battle, fighter, action, cost) {
    return action.isSkill(CHARACTER_ID, 1) ? cost - balance(fighter) : cost;
  },

  applyConditionEffects(battle, choice) {
    const actor = choice.actor;
    const action = choice.action;
    if (action.isSkill(CHARACTER_ID, 0)) {
      const bonus = balance(actor) * 2;
      if (bonus > 0) {
        choice.power = Number(choice.power || 0) + bonus;
        battle.logs.push(`${BALANCE} 중첩 수 ${balance(actor)}로 위력이 ${bonus} 증가했다.`);
      }
    }
    if (action.isSkill(CHARACTER_ID, 2) && balance(actor) < 1) return false;
    return undefined;
  },

  estimatedPower(_battle, actor, _target, action, power) {
    return action.isSkill(CHARACTER_ID, 0) ? power + balance(actor) * 2 : power;
  },

  attackDamageMultipliers(_battle, choice) {
    return choice.action.isAttack && Number(choice.actor.counters[PRAYER] || 0) > 0
      ? [1 + balance(choice.actor) * 0.2]
      : [];
  },

  estimatedDamageMultipliers(_battle, actor, _target, action) {
    return action.isAttack && Number(actor.counters[PRAYER] || 0) > 0 ? [1 + balance(actor) * 0.2] : [];
  },

  onHitAfterDefenseAsActor(battle, choice) {
    const actor = choice.actor;
    const target = battle.opponent(actor);
    if (choice.action.isSkill(CHARACTER_ID, 0)) {
      const [attacks, nonAttacks] = selectedCounts(target);
      battle.fixedDamage(target, Math.max(0, nonAttacks - attacks), choice.action.name, actor);
    } else if (choice.action.isSkill(CHARACTER_ID, 3)) {
      battle.fixedDamage(target, judgmentFixedDamage(target, balance(actor)), choice.action.name, actor);
    }
  },

  applyNonAttackEffects(battle, choice) {
    const actor = choice.actor;
    if (choice.action.isSkill(CHARACTER_ID, 1)) {
      battle.applyDefense(actor, choice.action.name);
      return true;
    }
    if (choice.action.isSkill(CHARACTER_ID, 2)) {
      actor.counters[PRAYER] = 2;
      battle.logs.push(`다음 턴 동안 공격 피해가 x${1 + balance(actor) * 0.2}가 된다.`);
      return true;
    }
    return false;
  },

  onDefenseHit(battle, choice) {
    const attacker = choice.actor;
    const defender = battle.opponent(attacker);
    if (defender.defenseName !== "지킨다는 것의 무거움") return;
    const [attacks, nonAttacks] = selectedCounts(attacker);
    battle.fixedDamage(attacker, Math.max(0, attacks - nonAttacks), defender.defenseName, defender);
  },

  onTurnEnd(battle, fighter) {
    if (battle.turn % 2 !== 0) return;
    const [attacks, nonAttacks] = selectedCounts(fighter);
    if (attacks === nonAttacks) addCounter(battle, fighter, BALANCE, 1);
  },

  decrementCounters(fighter) {
    if (!(PRAYER in fighter.counters)) return;
    const remaining = Number(fighter.counters[PRAYER] || 0);
    if (remaining <= 1) delete fighter.counters[PRAYER];
    else fighter.counters[PRAYER] = remaining - 1;
  },

  modifyFixedDamageToOpponent(battle, actor, _target, amount) {
    const bonus = balance(actor);
    if (bonus <= 0 || amount <= 0) return amount;
    battle.logs.push(`냉혹한 심판자: 고정 피해가 ${bonus} 증가했다.`);
    return amount + bonus;
  },

  setupValue(battle, actor, target, action) {
    const stacks = balance(actor);
    if (action.isSkill(CHARACTER_ID, 1)) {
      const incoming = battle.estimateBestIncomingDamage(target, actor);
      const [attacks, nonAttacks] = selectedCounts(target);
      return incoming * 0.25 + Math.max(0, attacks - nonAttacks) * 150 + stacks * 60;
    }
    if (action.isSkill(CHARACTER_ID, 2) && stacks >= 1) {
      if (Number(actor.counters[PRAYER] || 0) > 0) return 0;
      return 560 + stacks * 300;
    }
    return 0;
  },

  aiScore(battle, actor, target, action, expectedDamage, hitRate) {
    let value = 0;
    const stacks = balance(actor);
    const [ownAttacks, ownNonAttacks] = selectedCounts(actor);
    const [targetAttacks, targetNonAttacks] = selectedCounts(target);
    const incoming = battle.estimateBestIncomingDamage(target, actor);

    value += balanceTimingValue(battle, action, ownAttacks, ownNonAttacks);

    if (action.isSkill(CHARACTER_ID, 0)) {
      const extra = Math.max(0, targetNonAttacks - targetAttacks);
      value += stacks * 180 + extra * 260 * hitRate;
      if (extra > 0) value += 360;
    } else if (action.isSkill(CHARACTER_ID, 1)) {
      const punish = Math.max(0, targetAttacks - targetNonAttacks);
      const incomingWeight = targetAttacks > targetNonAttacks || incoming >= actor.hp * 0.75 ? 0.95 : 0.35;
      value += incoming * incomingWeight + punish * 330;
      if (incoming >= actor.hp) value += 2100;
      if (battle.turn % 2 === 0 && projectedEqual(action, ownAttacks, ownNonAttacks)) value += 460;
    } else if (action.isSkill(CHARACTER_ID, 2)) {
      if (stacks < 1) value -= 3200;
      else if (Number(actor.counters[PRAYER] || 0) > 0) value -= 900;
      else {
        value += 780 + stacks * 460;
        if (actor.mp >= JUDGMENT_RESERVE_MP + Math.floor(action.mp / 2)) value += 520;
      }
    } else if (action.isSkill(CHARACTER_ID, 3)) {
      const fixed = judgmentFixedDamage(target, stacks) + stacks;
      value += fixed * 2.8 * hitRate;
      if (expectedDamage + fixed * hitRate >= target.hp) value += 7200;
      else if (actor.mp < battle.effectiveCost(actor, action) + 18) value -= 520;
      if (Number(actor.counters[PRAYER] || 0) > 0) value += expectedDamage * 0.85;
    } else if (action.isCommonAction("meditation")) {
      if (actor.mp < 50) value += 360 + Math.max(0, 64 - actor.mp) * 9;
      if (battle.turn % 2 === 0 && projectedEqual(action, ownAttacks, ownNonAttacks)) value += 520;
      if (actor.mp >= 92) value -= 420;
    } else if (action.isCommonAction("normal_attack")) {
      if (stacks >= 2 && Number(actor.counters[PRAYER] || 0) > 0) value += expectedDamage * 0.7;
    }
    return value;
  },

  wouldConditionFail(_battle, actor, _target, action) {
    return action.isSkill(CHARACTER_ID, 2) ? balance(actor) < 1 : undefined;
  },
};
